### ISSUE TITLE
[java] New rules around Optional usage for Java 8+

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#1034]
I'm seeing some bad usages of Optional which I think we can translate into a bunch of new rules for Java.

1. Never return `null` in a method returning `Optional` / `OptionalInt` / `OptionalDouble` / `OptionalLong`
2. `if (optional.isPresent()) { single_line }` should be `optional.ifPresent(lambda)`
3. `optional.isPresent() ? Optional.of(something) : Optional.empty()` should be `optional.map(something_as_lambda)`
4. `optional.isPresent() ? something : Optional.empty()` should be `optional.flatMap(something_as_lambda)`

**Comment by oowekyala** [comment]
Additionally, with java 9+, 
```java
if (optional.isPresent()) {
  single_line1
} else {
  single_line2
}
```

can be replaced by `optional.ifPresentOrElse​(lambda1, lambda2)`

**Comment by andreybleme** [comment]
@jsotuyod  May I start to implement some of these rules? Are there any gitter I can get help if I need?

**Comment by jsotuyod** [comment]
@andreybleme most definitively!

**Comment by andreybleme** [comment]
@jsotuyod woudln't be better to have an issue for each rule? This way we could track the progress of each rule separately. i.e I'm starting to implement:

> Never return null in a method returning Optional / OptionalInt / OptionalDouble / OptionalLong. Use the empty() optionals instead.

**Comment by jsotuyod** [comment]
@andreybleme on this case, since the checks are very punctual, but all related to poor usage of the `Optional` APIs, so I'd rather have a single rule, and properties to enable / disable individual checks if it makes sense (ie: as we did with [`LinguisticNaming`](https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#linguisticnaming)

Nontheless, it's perfectly fine to build that rule incrementally, adding a check at a time.

**Comment by stefankatic** [comment]
Any progress on this?

**Comment by jsotuyod** [comment]
@stefankatic none, we haven't heard back from @andreybleme. Contributions are always welcomed.

**Comment by Fernal73** [comment]
My bad. I ought to have checked issues to see if there were any existing write-ups. 

I'll just post the article here too:

https://dzone.com/articles/using-optional-correctly-is-not-optional

**Comment by Fernal73** [comment]
It might be a good idea to revisit splitting the rule into multiple ones. That way, contributors can work independently on each piece.

**Comment by adangel** [comment]
The usage of `Optional.of(...)` vs. `Optional.ofNullable(...)` is also tricky. The first will throw a NPE, while the second will return an empty optional, if null was provided.

**Comment by Fernal73** [comment]
Simpler, easily understandable rules such as not allowing an Optional
object to be set to null and disallowing the use of Optionals as member
fields, in constructor , setter and method arguments , in collections and
returning empty collections or arrays as Optionals could be the first ones
to be implemented.

That's my take on it. Of course, that could be because I've yet to come to
terms with using Optional myself and thus I lean towards the rules that are
unambiguous enough to understand off the bat.
